shelley/chain-and-ledger/formal-spec/multi-sig.tex

```diff
@@ -0,0 +1,567 @@
+      \wcard)\in\txins{tx}\} \\~\\
+      \txttl tx \geq \var{slot}
+      & \txins{tx} \neq \emptyset
+      & \minfee{pp}{tx} \leq \txfee{tx}
+      & \var{txInps}\subseteq \dom \var{utxo}
+      \\
+      \consumed{pp}{utxo}{stkeys}{rewards}~{tx} = \produced{pp}{stpools}~{tx}
+      \\
+      ~
+      \\
+      \forall (\_\mapsto (\_, c)) \in \txouts{tx}, c \geq 0
+      \\
+      ~
+      \\
+      \var{refunded} \leteq \keyRefunds{pp}{stkeys}~{tx}
+      \\
+      \var{decayed} \leteq \decayedTx{pp}{stkeys}~{tx}
+      \\
+      \var{depositChange} \leteq
+        (\deposits{pp}~{stpools}~{\fun{dcerts}~tx}) - (\var{refunded} + \var{decayed})
+    }
+    {
+      \begin{array}{l}
+        \var{slot}\\
+        \var{pp}\\
+        \var{stkeys}\\
+        \var{stpools}\\
+      \end{array}
+      \vdash
+      \left(
+      \begin{array}{r}
+        \var{utxo} \\
+        \var{deposits} \\
+        \var{fees} \\
+      \end{array}
+      \right)
+      \trans{utxo}{tx}
+      \left(
+      \begin{array}{r}
+        \varUpdate{(\{(\var{txInps}\subtractdom \var{utxo}) \cup \outs{tx}}  \\
+        \varUpdate{\var{deposits} + \var{depositChange}} \\
+        \varUpdate{\var{fees} + \txfee{tx} + \var{decayed}} \\
+      \end{array}
+      \right)
+    }
+  \end{equation}
+  \caption{UTxO inference rules}
+  \label{fig:rules:utxo-mulit-sig}
+\end{figure}
+
+In order to work with the new types, some function definitions
+of~\cite{shelley_formal_spec} have to be extended. Those functions are listed in
+Figure~\ref{fig:functions-witnesses-multi-sig}. The function $\fun{witsNeeded}$
+only collects the payment keys for non-script addresses.
+
+The function $\fun{consumed}$ computes the sum of all coin values spent by a
+transaction. As the return value of the function $\fun{txins}$ has been changed,
+the parameter passed to $\fun{ubalance}$ has to be adapted, too, by projecting
+the returned triple onto the first two entries.
+
+\begin{figure}[htb]
+  \begin{align*}
+    & \fun{witsNeeded} \in \UTxO \to \Tx \to \powerset{\HashKey}
+    & \begin{array}{r}
+        \text{hashkeys of} \\
+        \text{required witnesses}
+      \end{array} \\
+    & \fun{witsNeeded}~{utxo}~{tx} = \\
+    & ~~\{ \fun{paymentHK}~a \mid i \mapsto (a, \wcard) \in
+      \var{utxo},~(i,\Nothing)\in\txins{tx}\}~\cup \\
+    & ~~\{\fun{stakeHK_r}~a \mid a\mapsto \wcard \in \txwdrls{tx}\}~\cup \\
+    & ~~\{\cwitness{c} \mid c \in \txcerts{tx}\}~\cup \\
+    & ~~\bigcup_{\substack{c \in \txcerts{tx} \\ ~c \in\DCertRegPool}}
+    \fun{poolOwners}~{c}
+    \nextdef
+    & \fun{consumed} \in \PParams \to \UTxO \to \StakeKeys \to \Wdrl \to \Tx \to \Coin
+    & \text{value consumed} \\
+    & \consumed{pp}{utxo}{stkeys}{rewards}~{tx} = \\
+    & ~~\ubalance{(\{(\var{txid},\var{ix}) \vert (\var{txid}, \var{ix},
+      \wcard)\in\txins{tx}\}
+      \restrictdom \var{utxo})} +
+      \fun{wbalance}~(\fun{txwdrls}~{tx}) \\
```

Comment:
the corners on this figure are broken, but I think if you move `+wbalance` to the next line it'll work.

Reply:
I re-arranged the last sum to fix this